resource_search derive klass

kbrock opened the pull request:
update resource_search in 2 (commits) ways:

1. Optionally pass a search key so searching by `name` instead of `id` does not need to reimplement `resource_search`.
2. Make `klass` optional for `resource_search` since it is basically just `collection_class(type)` anyway.

There are a number of `resource_search` instances that explicitly pass a different value into `collection_class` than `type`. I held off of updating these. They changes seemed to work fine but I'm trying to keep this PR as straight forward as possible.

kbrock commented on app/controllers/api/base_controller/generic.rb:
```diff
@@ -253,9 +251,9 @@ def model_ident(model, type = nil)
-        raise NotFoundError, "Invalid #{klass} id #{id} specified" unless id.kind_of?(Integer) || id =~ /\A\d+\z/
+      def validate_id(id, key_id, klass)
+        return if key_id != "id"
+        raise NotFoundError, "Invalid #{klass} id #{id} specified" unless id.kind_of?(Integer) || id.to_s =~ /\A\d+\z/
```
if id is nil, then this raises ruby warnings. `to_s` removes the warning and makes it always fail (which is what we want)

Fryguy replied:
No need to check for kind_of?(Integer) anymore, then

kbrock replied:
I was using to_s to fix the `nil`, but I guess it fixes the integers as well.

kbrock replied:
thanks for the pointer to `integer?`. Didn't realize rails overrode that for 'integer` and `string`.

Fryguy commented on app/controllers/api/automate_domains_controller.rb:
```diff
@@ -67,17 +67,8 @@ def automate_domain_ident(domain)
-        super(domain.id, type, klass)
-      end
+    def resource_search(id, type, klass = nil, key_id = nil)
+      super(id, type, klass, id.to_s =~ /\A\d+\z/ ? :id : :name)
```
```suggestion
      super(id, type, klass, id.to_s.integer? ? :id : :name)
```

Fryguy replied:
i see the old code was that way though, so no need to change for this PR.

kbrock replied:
done. thanks

kbrock commented:
update: integer instead of regex checks, swapped :id/:name to "id"/"name", changed validation so a nil name will throw better validation error.

Fryguy commented on app/controllers/api/generic_object_definitions_controller.rb:
```diff
@@ -136,17 +137,11 @@ def generic_objects_request?
-        go_def = klass.find_by!(:name => id)
-        filter_resource(go_def, type, klass)
-      end
+    def resource_search(id, type, klass = nil, _key_id = nil)
+      super(id, type, klass, id&.integer? ? "id" : "name")
     end
```
Wondering if this is clearer...up to you:

```suggestion
    def resource_search(id, type, klass = nil, key_id = nil)
      key_id = id&.integer? ? "id" : "name"
      super
    end
```

kbrock replied:
I link I like the old way but not a hill... done

Fryguy commented on app/controllers/api/service_orders_controller.rb:
```diff
@@ -26,13 +26,13 @@ def clear_resource(type, id, _data)
-    def validate_id(id, type, klass)
-      id == USER_CART_ID || super(id, type, klass)
+    def validate_id(id, key_id, klass)
+      id == USER_CART_ID || super(id, key_id, klass)
```
```suggestion
      id == USER_CART_ID || super
```

Fryguy commented on app/controllers/api/base_controller/generic.rb:
```diff
@@ -253,9 +251,10 @@ def model_ident(model, type = nil)
-        return if collection_config.resource_identifier(type) != "id"
-        raise NotFoundError, "Invalid #{klass} id #{id} specified" unless id.kind_of?(Integer) || id =~ /\A\d+\z/
+      def validate_id(id, key_id, klass)
+        if key_id == "id" && !id&.integer? || id.nil?
```
This is very confusing mixin && and || without the precendence...perhaps:

```suggestion
        if key_id == "id" && (id.nil? || !id.integer?)
```

kbrock replied:
the original code path went through with id if they specified nothing.
So it was not as lenient, and it properly stated that they did not have an id.

If we remove the 'id.nil?` then , then id.nil? case will go through with a name.
I did modify the validation to say that name requires a value as well, so the error should be thrown, but I wanted the text.

I'll try adding parens or something and see if it becomes more clear.

Fryguy commented on app/controllers/api/base_controller/generic.rb:
```diff
@@ -253,9 +251,10 @@ def model_ident(model, type = nil)
-        raise NotFoundError, "Invalid #{klass} id #{id} specified" unless id.kind_of?(Integer) || id =~ /\A\d+\z/
+      def validate_id(id, key_id, klass)
+        if key_id == "id" && !id&.integer? || id.nil?
+          raise NotFoundError, "Invalid #{klass} #{key_id} #{id} specified"
```
```suggestion
          raise NotFoundError, "Invalid #{klass} #{key_id} #{id.inspect} specified"
```

kbrock replied:
I didn't like the double quotes coming across, I did an `|| 'nil'`

Fryguy commented on app/controllers/api/base_controller/renderer.rb:
```diff
@@ -125,9 +125,10 @@ def virtual_attribute_accessor(type, attr)
-      def resource_search(id, type, klass)
-        validate_id(id, type, klass)
-        key_id = collection_config.resource_identifier(type)
+      def resource_search(id, type, klass = nil, key_id = nil)
+        klass  ||= collection_class(type)
+        key_id ||= collection_config.resource_identifier(type)
```
Instead of making key_id a parameter, which is derived in every case except 1, perhaps we should make it a subclassable method, similar to validate_id, which is also only overridden in one case.  In fact, maybe it makes sense to move the "key_id" concept into validate_id, perhaps having that method _return_ the `key_id`

kbrock replied:
Yea, I think we only have 2 or 3 cases where we pass in a string or an id.

I liked passing in id here because the method that calls this has the knowledge that a string name is being used to look up. So it consolidates the logic.

I like the idea of deferring this decision.

kbrock replied:
The over arching goal is to set this up in a way that `api_resource` can just call `resource_search` without any special cases and it would work.

Currently, `resource_search` for these cases require `data['name']`. This is not available so that is why there is all this kludging.

Maybe once klass (and possibly key_id) is gone, then that will open us up for passing in the `data` or `name` parameter and they will all fall into the common pattern.

kbrock commented:
I'm being indecisive:

Currently, we swap the id passed into search and then override the search to handle that change.
It seems to make more sense to have the searching method state the intent in the first place.

```ruby
    def fetch_generic_object_definition(type, id, data)
      # current code
      resource_search(id || data['name'], type)
      # a
      id&.integer? ? resource_search(id, type) : resource_search(data['name'], type, nil, "name")
      # b
      resource_search(id || data['name'], type, nil, id ? nil : "name")
      # c
      if id.nil?
        resource_search(data['name'], type, nil, "name")
      else
        resource_search(id, type)
      end
    end

    # only needed for first case
    def resource_search(id, type, klass = nil, key_id = nil)
      key_id = "name" unless id&.integer?
      super
    end

kbrock commented:
push:
- changed NotFound to BadRequest when the id is wrong, and changed the accompanying specs
- used `super` when overriding search behavior from id to name
